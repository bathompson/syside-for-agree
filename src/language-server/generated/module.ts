/********************************************************************************
 * Copyright (c) 2022-2023 Sensmetry UAB and others
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is
 * available at https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/

/******************************************************************************
 * This file was generated by langium-cli 0.5.0.
 * DO NOT EDIT MANUALLY!
 ******************************************************************************/

import { LangiumGeneratedServices, LangiumGeneratedSharedServices, LangiumSharedServices, LangiumServices, LanguageMetaData, Module, IParserConfig } from 'langium';
import { SysMlAstReflection } from './ast';
import { KerMLGrammar, SysMLGrammar } from './grammar';

export const KerMLLanguageMetaData: LanguageMetaData = {
    languageId: 'kerml',
    fileExtensions: ['.kerml'],
    caseInsensitive: false
};

export const SysMLLanguageMetaData: LanguageMetaData = {
    languageId: 'sysml',
    fileExtensions: ['.sysml'],
    caseInsensitive: false
};

export const parserConfig: IParserConfig = {
    recoveryEnabled: true,
    nodeLocationTracking: 'none',
};

export const SysMlGeneratedSharedModule: Module<LangiumSharedServices, LangiumGeneratedSharedServices> = {
    AstReflection: () => new SysMlAstReflection()
};

export const KerMLGeneratedModule: Module<LangiumServices, LangiumGeneratedServices> = {
    Grammar: () => KerMLGrammar(),
    LanguageMetaData: () => KerMLLanguageMetaData,
    parser: {
        ParserConfig: () => parserConfig
    }
};

export const SysMLGeneratedModule: Module<LangiumServices, LangiumGeneratedServices> = {
    Grammar: () => SysMLGrammar(),
    LanguageMetaData: () => SysMLLanguageMetaData,
    parser: {
        ParserConfig: () => parserConfig
    }
};
