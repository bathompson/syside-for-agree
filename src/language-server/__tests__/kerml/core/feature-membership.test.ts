/********************************************************************************
 * Copyright (c) 2022-2023 Sensmetry UAB and others
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License, v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is
 * available at https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/

import { anything, withQualifiedName, qualifiedTypeReference } from "../../../../testing";
import { Classifier, Feature, MultiplicityRange, LiteralInfinity } from "../../../generated/ast";

test("features can be parsed and aliased", async () => {
    return expect(`
    datatype Integer;
    feature person[*] : Person;
    classifier Person {
        feature age[1]: Integer;
        alias personAlias for person;
    }`).toParseKerML({
        elements: [
            ...anything(1),

            {
                $type: Classifier,
                ...withQualifiedName("Person"),
                features: [
                    {
                        $type: Feature,
                        ...withQualifiedName("Person::age"),
                        typedBy: [qualifiedTypeReference("Integer")],
                    },
                ],
                aliases: [
                    {
                        ...withQualifiedName("Person::personAlias"),
                        for: qualifiedTypeReference("person"),
                    },
                ],
            },
        ],
        features: [
            {
                $type: Feature,
                ...withQualifiedName("person"),
                typedBy: [qualifiedTypeReference("Person")],
                multiplicity: { $type: MultiplicityRange, range: { $type: LiteralInfinity } },
            },
        ],
    });
});

test("member features can be parsed", async () => {
    return expect(`
    classifier A;
    classifier B {
        feature f;
        member feature g featured by A;
    }`).toParseKerML({
        elements: [
            ...anything(1),
            {
                $type: Classifier,
                features: [
                    {
                        $type: Feature,
                        ...withQualifiedName("B::f"),
                    },
                ],
                members: [
                    {
                        $type: Feature,
                        ...withQualifiedName("B::g"),
                        featuredBy: [qualifiedTypeReference("A")],
                    },
                ],
            },
        ],
    });
});
