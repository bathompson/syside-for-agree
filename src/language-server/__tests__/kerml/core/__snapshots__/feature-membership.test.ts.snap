// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features can be parsed and aliased: toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [],
    "filters": [],
    "imports": [],
    "members": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Feature",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "person",
          "filters": [],
          "imports": [],
          "isNonunique": false,
          "isOrdered": false,
          "isSufficient": false,
          "members": [],
          "multiplicity": {
            "$type": "OwningMembership",
            "annotations": [],
            "chains": [],
            "element": {
              "$type": "MultiplicityRange",
              "aliases": [],
              "annotatingMembers": [],
              "filters": [],
              "imports": [],
              "isNonunique": false,
              "isOrdered": false,
              "isSufficient": false,
              "members": [],
              "namespaceMembers": [],
              "prefixes": [],
              "range": {
                "$type": "OwningMembership",
                "annotations": [],
                "chains": [],
                "element": {
                  "$type": "LiteralInfinity",
                  "aliases": [],
                  "annotatingMembers": [],
                  "filters": [],
                  "imports": [],
                  "isNonunique": false,
                  "isOrdered": false,
                  "isSufficient": false,
                  "members": [],
                  "namespaceMembers": [],
                  "prefixes": [],
                  "relationshipMembers": [],
                  "typeRelationships": [],
                },
                "elements": [],
                "prefixes": [],
              },
              "relationshipMembers": [],
              "typeRelationships": [],
            },
            "elements": [],
            "prefixes": [],
          },
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [
            {
              "$type": "FeatureTyping",
              "annotations": [],
              "chains": [],
              "elements": [],
              "prefixes": [],
              "reference": {
                "$type": "TypeReference",
                "parts": [
                  {
                    "error": undefined,
                    "ref": {
                      "$type": "Classifier",
                      "aliases": [
                        {
                          "$type": "Membership",
                          "annotations": [],
                          "chains": [],
                          "declaredName": "personAlias",
                          "elements": [],
                          "prefixes": [],
                          "reference": {
                            "$type": "ElementReference",
                            "parts": [
                              {
                                "error": undefined,
                                "ref": [Circular],
                              },
                            ],
                          },
                        },
                      ],
                      "annotatingMembers": [],
                      "declaredName": "Person",
                      "filters": [],
                      "imports": [],
                      "isSufficient": false,
                      "members": [
                        {
                          "$type": "FeatureMembership",
                          "annotations": [],
                          "chains": [],
                          "element": {
                            "$type": "Feature",
                            "aliases": [],
                            "annotatingMembers": [],
                            "declaredName": "age",
                            "filters": [],
                            "imports": [],
                            "isNonunique": false,
                            "isOrdered": false,
                            "isSufficient": false,
                            "members": [],
                            "multiplicity": {
                              "$type": "OwningMembership",
                              "annotations": [],
                              "chains": [],
                              "element": {
                                "$type": "MultiplicityRange",
                                "aliases": [],
                                "annotatingMembers": [],
                                "filters": [],
                                "imports": [],
                                "isNonunique": false,
                                "isOrdered": false,
                                "isSufficient": false,
                                "members": [],
                                "namespaceMembers": [],
                                "prefixes": [],
                                "range": {
                                  "$type": "OwningMembership",
                                  "annotations": [],
                                  "chains": [],
                                  "element": {
                                    "$type": "LiteralNumber",
                                    "aliases": [],
                                    "annotatingMembers": [],
                                    "filters": [],
                                    "imports": [],
                                    "isNonunique": false,
                                    "isOrdered": false,
                                    "isSufficient": false,
                                    "literal": 1,
                                    "members": [],
                                    "namespaceMembers": [],
                                    "prefixes": [],
                                    "relationshipMembers": [],
                                    "typeRelationships": [],
                                  },
                                  "elements": [],
                                  "prefixes": [],
                                },
                                "relationshipMembers": [],
                                "typeRelationships": [],
                              },
                              "elements": [],
                              "prefixes": [],
                            },
                            "namespaceMembers": [],
                            "prefixes": [],
                            "relationshipMembers": [],
                            "typeRelationships": [
                              {
                                "$type": "FeatureTyping",
                                "annotations": [],
                                "chains": [],
                                "elements": [],
                                "prefixes": [],
                                "reference": {
                                  "$type": "TypeReference",
                                  "parts": [
                                    {
                                      "error": undefined,
                                      "ref": {
                                        "$type": "DataType",
                                        "aliases": [],
                                        "annotatingMembers": [],
                                        "declaredName": "Integer",
                                        "filters": [],
                                        "imports": [],
                                        "isSufficient": false,
                                        "members": [],
                                        "namespaceMembers": [],
                                        "prefixes": [],
                                        "relationshipMembers": [],
                                        "typeRelationships": [],
                                      },
                                    },
                                  ],
                                },
                              },
                            ],
                          },
                          "elements": [],
                          "prefixes": [],
                        },
                      ],
                      "namespaceMembers": [],
                      "prefixes": [],
                      "relationshipMembers": [],
                      "typeRelationships": [],
                    },
                  },
                ],
              },
            },
          ],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "DataType",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "Integer",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Classifier",
          "aliases": [
            {
              "$type": "Membership",
              "annotations": [],
              "chains": [],
              "declaredName": "personAlias",
              "elements": [],
              "prefixes": [],
              "reference": {
                "$type": "ElementReference",
                "parts": [
                  {
                    "error": undefined,
                    "ref": {
                      "$type": "Feature",
                      "aliases": [],
                      "annotatingMembers": [],
                      "declaredName": "person",
                      "filters": [],
                      "imports": [],
                      "isNonunique": false,
                      "isOrdered": false,
                      "isSufficient": false,
                      "members": [],
                      "multiplicity": {
                        "$type": "OwningMembership",
                        "annotations": [],
                        "chains": [],
                        "element": {
                          "$type": "MultiplicityRange",
                          "aliases": [],
                          "annotatingMembers": [],
                          "filters": [],
                          "imports": [],
                          "isNonunique": false,
                          "isOrdered": false,
                          "isSufficient": false,
                          "members": [],
                          "namespaceMembers": [],
                          "prefixes": [],
                          "range": {
                            "$type": "OwningMembership",
                            "annotations": [],
                            "chains": [],
                            "element": {
                              "$type": "LiteralInfinity",
                              "aliases": [],
                              "annotatingMembers": [],
                              "filters": [],
                              "imports": [],
                              "isNonunique": false,
                              "isOrdered": false,
                              "isSufficient": false,
                              "members": [],
                              "namespaceMembers": [],
                              "prefixes": [],
                              "relationshipMembers": [],
                              "typeRelationships": [],
                            },
                            "elements": [],
                            "prefixes": [],
                          },
                          "relationshipMembers": [],
                          "typeRelationships": [],
                        },
                        "elements": [],
                        "prefixes": [],
                      },
                      "namespaceMembers": [],
                      "prefixes": [],
                      "relationshipMembers": [],
                      "typeRelationships": [
                        {
                          "$type": "FeatureTyping",
                          "annotations": [],
                          "chains": [],
                          "elements": [],
                          "prefixes": [],
                          "reference": {
                            "$type": "TypeReference",
                            "parts": [
                              {
                                "error": undefined,
                                "ref": [Circular],
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                ],
              },
            },
          ],
          "annotatingMembers": [],
          "declaredName": "Person",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [
            {
              "$type": "FeatureMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Feature",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "age",
                "filters": [],
                "imports": [],
                "isNonunique": false,
                "isOrdered": false,
                "isSufficient": false,
                "members": [],
                "multiplicity": {
                  "$type": "OwningMembership",
                  "annotations": [],
                  "chains": [],
                  "element": {
                    "$type": "MultiplicityRange",
                    "aliases": [],
                    "annotatingMembers": [],
                    "filters": [],
                    "imports": [],
                    "isNonunique": false,
                    "isOrdered": false,
                    "isSufficient": false,
                    "members": [],
                    "namespaceMembers": [],
                    "prefixes": [],
                    "range": {
                      "$type": "OwningMembership",
                      "annotations": [],
                      "chains": [],
                      "element": {
                        "$type": "LiteralNumber",
                        "aliases": [],
                        "annotatingMembers": [],
                        "filters": [],
                        "imports": [],
                        "isNonunique": false,
                        "isOrdered": false,
                        "isSufficient": false,
                        "literal": 1,
                        "members": [],
                        "namespaceMembers": [],
                        "prefixes": [],
                        "relationshipMembers": [],
                        "typeRelationships": [],
                      },
                      "elements": [],
                      "prefixes": [],
                    },
                    "relationshipMembers": [],
                    "typeRelationships": [],
                  },
                  "elements": [],
                  "prefixes": [],
                },
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [
                  {
                    "$type": "FeatureTyping",
                    "annotations": [],
                    "chains": [],
                    "elements": [],
                    "prefixes": [],
                    "reference": {
                      "$type": "TypeReference",
                      "parts": [
                        {
                          "error": undefined,
                          "ref": {
                            "$type": "DataType",
                            "aliases": [],
                            "annotatingMembers": [],
                            "declaredName": "Integer",
                            "filters": [],
                            "imports": [],
                            "isSufficient": false,
                            "members": [],
                            "namespaceMembers": [],
                            "prefixes": [],
                            "relationshipMembers": [],
                            "typeRelationships": [],
                          },
                        },
                      ],
                    },
                  },
                ],
              },
              "elements": [],
              "prefixes": [],
            },
          ],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [],
  },
}
`;

exports[`member features can be parsed: toParseKerML 1`] = `
{
  "diagnostics": [],
  "lexerErrors": [],
  "parserErrors": [],
  "value": {
    "$type": "Namespace",
    "aliases": [],
    "annotatingMembers": [],
    "filters": [],
    "imports": [],
    "members": [],
    "namespaceMembers": [
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Classifier",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "A",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
      {
        "$type": "OwningMembership",
        "annotations": [],
        "chains": [],
        "element": {
          "$type": "Classifier",
          "aliases": [],
          "annotatingMembers": [],
          "declaredName": "B",
          "filters": [],
          "imports": [],
          "isSufficient": false,
          "members": [
            {
              "$type": "FeatureMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Feature",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "f",
                "filters": [],
                "imports": [],
                "isNonunique": false,
                "isOrdered": false,
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [],
              },
              "elements": [],
              "prefixes": [],
            },
            {
              "$type": "OwningMembership",
              "annotations": [],
              "chains": [],
              "element": {
                "$type": "Feature",
                "aliases": [],
                "annotatingMembers": [],
                "declaredName": "g",
                "filters": [],
                "imports": [],
                "isNonunique": false,
                "isOrdered": false,
                "isSufficient": false,
                "members": [],
                "namespaceMembers": [],
                "prefixes": [],
                "relationshipMembers": [],
                "typeRelationships": [
                  {
                    "$type": "TypeFeaturing",
                    "annotations": [],
                    "chains": [],
                    "elements": [],
                    "prefixes": [],
                    "reference": {
                      "$type": "TypeReference",
                      "parts": [
                        {
                          "error": undefined,
                          "ref": {
                            "$type": "Classifier",
                            "aliases": [],
                            "annotatingMembers": [],
                            "declaredName": "A",
                            "filters": [],
                            "imports": [],
                            "isSufficient": false,
                            "members": [],
                            "namespaceMembers": [],
                            "prefixes": [],
                            "relationshipMembers": [],
                            "typeRelationships": [],
                          },
                        },
                      ],
                    },
                  },
                ],
              },
              "elements": [],
              "prefixes": [],
            },
          ],
          "namespaceMembers": [],
          "prefixes": [],
          "relationshipMembers": [],
          "typeRelationships": [],
        },
        "elements": [],
        "prefixes": [],
      },
    ],
    "prefixes": [],
    "relationshipMembers": [],
  },
}
`;
